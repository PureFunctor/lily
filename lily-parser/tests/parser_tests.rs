use lily_parser::parse_top_level;

#[test]
pub fn top_level_0() {
    let source = r"
infixl 4 add as +

example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_1() {
    let source = r"
example = a + b + c
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_2() {
    let source = r"
infixl 1 add as +
infixl 2 add as *

example = a + b * (c + d) + e
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_3() {
    let source = r"
infixl 1 add as +

example = if if a then b else c then if d then e else f else if g then h else i
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_4() {
    let source = r"
example = f a b
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}

#[test]
pub fn top_level_5() {
    let source = r"
example = f if a then b else c d

example = f (if a then b else c) d
";

    insta::assert_debug_snapshot!(parse_top_level(source));
}
