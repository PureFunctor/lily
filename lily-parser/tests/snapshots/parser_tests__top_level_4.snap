---
source: lily-parser/tests/parser_tests.rs
expression: parse_top_level(source)
---
Ok(
    Module {
        declarations: [
            Declaration {
                begin: 1,
                end: 16,
                kind: ValueDeclaration(
                    "example",
                    [],
                    Expression {
                        begin: 11,
                        end: 16,
                        kind: Application(
                            Expression {
                                begin: 11,
                                end: 12,
                                kind: Variable(
                                    "f",
                                ),
                            },
                            [
                                Expression {
                                    begin: 13,
                                    end: 14,
                                    kind: Variable(
                                        "a",
                                    ),
                                },
                                Expression {
                                    begin: 15,
                                    end: 16,
                                    kind: Variable(
                                        "b",
                                    ),
                                },
                            ],
                        ),
                    },
                ),
            },
        ],
    },
)
